pytorch_lightning/plugins/ddp_plugin.py

```diff
@@ -91,23 +92,23 @@ def init_ddp_connection(
                 torch_backend, rank=global_rank, world_size=world_size
             )
 
-    def on_before_forward(self, model: LightningModule, *args):
+    def on_before_forward(self, model: LightningModule, *batch):
         """
         Override to handle custom input to device logic. For DDP, no logic is required as this is handled internally
         within the DDP wrapper.
 
         Example::
 
-            def on_before_forward(self, model, *args):
-                batch, batch_idx = args
+            def on_before_forward(self, model, *batch):
+                batch, batch_idx = batch
                 return batch.to(model.device)
```

Comment:
input is `batch, batch_idx`, but it returns only `batch` in this example?

Reply:
also batch_transfer to a particular device is possible with DDP? isn't this handled by pytorch?

Reply:
It doesn't seem to be handled for custom batches.

Reply:
so for normal batches (tensors..), this still works if we put the batches on the correct device before pytorch does in https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html??

Reply:
@rohitgr7 this DDP from pytorch is restricted to a few primitives (tensors and dict I believe). DDP actually does more than we  use it for here in Lightning, it supports multiple device indices for which a batch would have to be scattered. For non-tensor objects, generally, the batch size (and batch dim) is undefined therefore scattering is not possible (and gathering for that matter). However, in Lightning we never pass multiple device indices to DDP. Instead, we launch multiple processes and in each we pass a single device indes. This usage is also recommended by pytorch. It also means we don't have to scatter any batches. This is how I understand it, hope it is clearer. 

For DP the issue remains, because there it's always required to scatter and gather (multiple devices).

Reply:
ah.. got it now.. thanks @awaelchli . Greate explanation